src/graph/validation/invariant_checks.rs

```diff
@@ -0,0 +1,148 @@
+// License: MIT
+// Copyright © 2024 Frequenz Energy-as-a-Service GmbH
+
+//! Helper methods for checking invariants of a [`ComponentGraph`].
+
+use crate::{ComponentCategory, Edge, Error, Node};
+
+use super::ComponentGraphValidator;
+
+impl<N, E> ComponentGraphValidator<'_, N, E>
+where
+    N: Node,
+    E: Edge,
+{
+    /// Checks that the given node is a leaf node.
+    pub(super) fn ensure_leaf(&self, node: &N) -> Result<(), Error> {
+        if let Some(successor) = self.cg.successors(node.component_id())?.next() {
+            return Err(Error::invalid_graph(format!(
+                "{}:{} can't have any successors. Found {}:{}.",
+                node.category(),
+                node.component_id(),
+                successor.category(),
+                successor.component_id()
+            )));
+        }
+        Ok(())
+    }
+
+    /// Checks that the given node is *not* a leaf node.
+    pub(super) fn ensure_not_leaf(&self, node: &N) -> Result<(), Error> {
+        if self.cg.successors(node.component_id())?.next().is_none() {
+            return Err(Error::invalid_graph(format!(
+                "{}:{} must have at least one successor.",
+                node.category(),
+                node.component_id()
+            )));
+        }
+        Ok(())
+    }
+
+    /// Checks that the given node is a root node.
+    pub(super) fn ensure_root(&self, node: &N) -> Result<(), Error> {
+        if let Some(predecessor) = self.cg.predecessors(node.component_id())?.next() {
+            return Err(Error::invalid_graph(format!(
+                "{}:{} can't have any predecessors. Found {}:{}.",
+                node.category(),
+                node.component_id(),
+                predecessor.category(),
+                predecessor.component_id()
+            )));
+        }
+        Ok(())
+    }
+
+    /// Checks that the given node only has predecessors with the given categories.
+    pub(super) fn ensure_predecessor_categories(
```

Comment:
Wouldn't it make sense to generalize this to ensure a predecessor (or successor) fulfills some predicate (function)?

Reply:
that makes sense, especially when we want to introduce BatteryType, etc.

done.